### PR TITLE
Refactor/categories spec

### LINKED ISSUE #1822: Refactor spec/controllers/categories_controller_spec.rb to spec/request/category_spec.rb
# Description

Controller Specs had been obsolete since Rails 5. It is best practice to create request spec instead of controller spec as recommended by the Rspec team.

Recommended Reading:

- [Stackoverflow Discussion in regards to Controller Specs vs Request Specs](https://stackoverflow.com/questions/40851705/controller-specs-vs-request-specs)
- [Medium article discussing request and controller specs](https://medium.com/just-tech/rspec-controller-or-request-specs-d93ef563ef11)

Resources on Request Spec formatting:

- [RelishApp](https://relishapp.com/rspec/rspec-rails/v/4-0/docs/request-specs/request-spec)

Original Issue: https://github.com/ifmeorg/ifme/issues/1815

---

Please assign yourself (via the Assignees dropdown), if you do want to work on this issue. Can't find yourself? You need to [join our organization](https://github.com/ifmeorg/ifme/wiki/Join-Our-Slack).

Check out our [Picking Up Issues](https://github.com/ifmeorg/ifme/wiki/Picking-Up-Issues) guide if you haven't already!

### PR DESCRIPTION
**Pull request by andrewhchang**
# Description
Refactor Categories controller spec to a request spec

## Corresponding Issue
closes #1822

### CODE REVIEW

**Comment by andrewhchang** [comment]
I'm wondering if there's any way to make git diff the two files? :\

**Comment by julianguyen** [comment]
Did you delete the file and then create a new one? I think if you do `git mv [old file] [new one]` you end up with a diff.

**Review by julianguyen — DISMISSED** [review]
Solid work! Thanks for taking this on 🎊

**Review by bennpham — APPROVED** [review]
This looks fine to me at the moment. 👍 

I'm reading through this comment http://rspec.info/blog/2016/07/rspec-3-5-has-been-released/#comment-3868711294 in regards to Rspec recommendation for request spec over controller specs and I'm wondering to myself whether `render_template` matcher is part of the `rails-controller-testing` gem or not. If it is, then we'd probably want to switch that to `expect(response).to be_successful` instead. If its still fine, I think it should do. Ideally, I think once all of the controller specs are converted to request specs, we should be able to remove the `rails-controller-testing` gem.

I got to head off for a bit so I haven't check it myself to see yet, but I might take a look once I'm back on my computer.

File: spec/requests/categories_spec.rb [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        headers = { 'ACCEPT' => 'application/json' }
+        get categories_path, headers: headers, params: params
+        expect(JSON.parse(response.body)['data'].first['name']).to eq(
+          category.name
+        )
+      end
+
+      it 'renders the page' do
+        get categories_path
+        expect(response).to render_template('index')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:22]:
```suggestion
        expect(response).to be_successful
```

File: spec/requests/categories_spec.rb [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+  describe '#show' do
+    context 'when the user is logged in' do
+      before { sign_in user }
+
+      context 'when the user created the category' do
+        before { get category_path(category.id) }
+
+        it 'renders the page' do
+          expect(response).to render_template('show')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:40]:
```suggestion
          expect(response).to be_successful
```

File: spec/requests/categories_spec.rb [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+
+  describe '#new' do
+    context 'when the user is logged in' do
+      before do
+        sign_in user
+        get new_category_path
+      end
+
+      it 'renders the page' do
+        expect(response).to render_template('new')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:68]:
```suggestion
        expect(response).to be_successful
```

File: spec/requests/categories_spec.rb [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+
+  describe '#edit' do
+    context 'when the user is logged in' do
+      before { sign_in user }
+
+      context 'user is trying to edit a category they created' do
+        before { get edit_category_path(category.id) }
+
+        it 'renders the edit form' do
+          expect(response).to render_template('edit')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:87]:
```suggestion
          expect(response).to be_successful
```

File: spec/requests/categories_spec.rb [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+        end
+      end
+
+      context 'when invalid params are supplied' do
+        let(:invalid_category_params) { { name: nil, description: nil } }
+        before do
+          post categories_path, params: { category: invalid_category_params }
+        end
+        it 're-renders the creation form' do
+          expect(response).to render_template('new')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:139]:
```suggestion
          expect(response).to be_successful
```

File: spec/requests/categories_spec.rb [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+      context 'when invalid params are supplied' do
+        let(:invalid_update_params) { { name: nil } }
+
+        before do
+          patch category_path(category),
+                params: { id: category.id, category: invalid_update_params }
+        end
+
+        it 're-renders the edit form' do
+          expect(response).to render_template('edit')
```
Comment (bennpham) [inline spec/requests/categories_spec.rb:219]:
```suggestion
          expect(response).to be_successful
```

**Review by bennpham — CHANGES_REQUESTED** [review]
Let's change all these `render_template` to `be_successful`. I just checked by removing `rails-controller-testing` gem and request spec don't support `render_template` by default.

It is [recommended to avoid the gem](http://rspec.info/blog/2016/07/rspec-3-5-has-been-released/#comment-3869319771) so if you assert that going to this URL is successful, that should be sufficient 👍

**Comment by andrewhchang** [comment]
> Did you delete the file and then create a new one? I think if you do `git mv [old file] [new one]` you end up with a diff.

yeah I did, I'll use git mv` and see if i can merge the two files so that I keep the commits

**Comment by andrewhchang** [comment]
> Let's change all these `render_template` to `be_successful`. I just checked by removing `rails-controller-testing` gem and request spec don't support `render_template` by default.
> 
> It is [recommended to avoid the gem](http://rspec.info/blog/2016/07/rspec-3-5-has-been-released/#comment-3869319771) so if you assert that going to this URL is successful, that should be sufficient 👍

Sure thing! Happy to :)

**Comment by andrewhchang** [comment]
not showing a diff :( but I preserved the old commits

**Review by bennpham — APPROVED** [review]
Looks good 👍

**Review by julianguyen — APPROVED** [review]
Great work! Feel free to merge!

**Pull request merged**